kore/src/Kore/Step/Simplification/AndTerms.hs

```diff
@@ -740,37 +785,28 @@ This unification case throws an error because domain values may not occur in a
 sort with constructors.
 -}
 domainValueAndConstructorErrors ::
-    Monad unifier =>
     HasCallStack =>
-    TermLike RewritingVariableName ->
-    TermLike RewritingVariableName ->
-    MaybeT unifier a
-domainValueAndConstructorErrors
-    term1@(DV_ _ _)
-    term2@(App_ secondHead _)
-        | Symbol.isConstructor secondHead =
-            error
-                ( unlines
-                    [ "Cannot handle DomainValue and Constructor:"
-                    , unparseToString term1
-                    , unparseToString term2
-                    ]
-                )
-domainValueAndConstructorErrors
-    term1@(App_ firstHead _)
-    term2@(DV_ _ _)
-        | Symbol.isConstructor firstHead =
-            error
-                ( unlines
-                    [ "Cannot handle Constructor and DomainValue:"
-                    , unparseToString term1
-                    , unparseToString term2
-                    ]
-                )
-domainValueAndConstructorErrors _ _ = empty
+    DVConstrError ->
+    unifier a
+domainValueAndConstructorErrors unifyData =
+    error $
+        show
+            ( Pretty.vsep
+                [ cannotHandle
+                , fromString $ unparseToString term1
+                , fromString $ unparseToString term2
+                , ""
+                ]
+            )
+  where
+    (term1, term2, cannotHandle) =
+        case unifyData of
+            DVConstr a b -> (a, b, "Cannot handle DomainValue and Constructor:")
+            ConstrDV a b -> (a, b, "Cannot handle Constructor and DomainValue:")
 
 data UnifyDomainValue = UnifyDomainValue
     { val1, val2 :: !(TermLike RewritingVariableName)
+    , term1, term2 :: !(TermLike RewritingVariableName)
```

Comment:
Similarly, it looks like we keeping the original terms just for sending to `explainBottom`. I think we can just transform `val1` and `val2` to back to `TermLike`s for that purpose.